File: testing/MLDB-466-full-reddit.js
// This file is part of MLDB. Copyright 2015 Datacratic. All rights reserved.

/* Example script to import a reddit dataset and run an example */

function assertEqual(expr, val, msg)
{
    if (expr == val)
        return;
    if (JSON.stringify(expr) == JSON.stringify(val))
        return;

    plugin.log("expected", val);
    plugin.log("received", expr);

    throw "Assertion failure: " + msg + ": " + JSON.stringify(expr)
        + " not equal to " + JSON.stringify(val);
}

function succeeded(response)
{
    return response.responseCode >= 200 && response.responseCode < 400;
}

function assertSucceeded(process, response)
{
    plugin.log(process, response.json);

    if (!succeeded(response)) {
        throw process + " failed: " + JSON.stringify(response);
    }
}

res = mldb.put('/v1/procedures/import_reddit', { 
    "type": "import.text",  
    "params": { 
        "dataFileUrl": "http://public.mldb.ai/reddit.csv.gz",
        'delimiter':'', 
        'quotechar':'',
        'outputDataset': 'reddit_raw',
        'runOnCreation': true,
        'limit': 10000
    } 
});

assertSucceeded("import text", res);

res = mldb.query("select * from reddit_raw limit 5");

mldb.log(res);

res = mldb.put('/v1/procedures/reddit_import', {
    "type": "transform",
    "params": {
        "inputData": "select tokenize(lineText, {offset: 1, value: 1}) as * from reddit_raw",
        "outputDataset": "reddit_dataset",
        "runOnCreation": true
    }
});

assertSucceeded("tokenize", res);

res = mldb.put('/v1/procedures/reddit_count_users', {
    "type": "transform",
    "params": {
        "inputData": "select columnCount() as numUsers named rowName() + '|1' from transpose(reddit_dataset)",
        "outputDataset": "reddit_user_counts",
        "runOnCreation": true
    }
});

assertSucceeded("user counts", res);

res = mldb.query("select * from reddit_dataset limit 5");

mldb.log(res);

res = mldb.put('/v1/procedures/reddit_svd', {
    "type" : "svd.train",
    "params" : {
        "trainingData" : "SELECT COLUMN EXPR (AS columnName() ORDER BY rowCount() DESC, columnName() LIMIT 1000) FROM reddit_dataset",
        "columnOutputDataset" : "reddit_svd_embedding",
        "runOnCreation": true
    }
});

assertSucceeded("svd", res);

res = mldb.query("select * from reddit_svd_embedding limit 5");

mldb.log(res);

res = mldb.put('/v1/procedures/reddit_kmeans', {
    "type" : "kmeans.train",
    "params" : {
        "trainingData" : "select * from reddit_svd_embedding",
        "outputDataset" : "reddit_kmeans_clusters",
        "numClusters" : 20,
        "runOnCreation": true
    }
});

assertSucceeded("kmeans", res);

res = mldb.query("select * from reddit_kmeans_clusters limit 5");

mldb.log(res);

res = mldb.put('/v1/procedures/reddit_tsne', {
    "type" : "tsne.train",
    "params" : {
        "trainingData" : "select * from reddit_svd_embedding",
        "rowOutputDataset" : "reddit_tsne_embedding",
        "runOnCreation": true
    }
});

assertSucceeded("tsne", res);

mldb.log(res.json);

res = mldb.query("select * from reddit_tsne_embedding limit 5");

mldb.log(res);

res = mldb.query("select * from reddit_user_counts limit 5");

mldb.log(res);

res = mldb.query(
    "select *, quantize(x, 7) as grid_x, quantize(y, 7) as grid_y " +
    "named regex_replace(rowName(), '\|1', '') " +
    "from merge(reddit_user_counts, reddit_tsne_embedding, reddit_kmeans_clusters) " +
    "where cluster is not null " +
    "order by numUsers desc limit 100"
);

mldb.log(res);

var mergedConfig = {
    type: "merged",
    id: "reddit_merged",
    params: {
        datasets: [
            { id: 'reddit_kmeans_clusters' },
            { id: 'reddit_tsne_embedding' },
            { id: 'reddit_user_counts' }
        ]
    }
};

var merged = mldb.createDataset(mergedConfig);

plugin.log(mldb.get("/v1/datasets/reddit_merged/query", {select:'*', limit:100}).json);

//MLDB-1176 merge function in FROM expression

var expected = mldb.get("/v1/query", {q:"SELECT * FROM reddit_merged LIMIT 10", format:'table'});
plugin.log(expected)

var resp = mldb.get("/v1/query", {q:"SELECT * FROM merge(reddit_kmeans_clusters, reddit_tsne_embedding, reddit_user_counts) LIMIT 10", format:'table'});
plugin.log(resp)

assertEqual(resp.json, expected.json);

"success"

File: testing/MLDB-466-full-reddit.py
#
# This file is part of MLDB. Copyright 2016 mldb.ai inc. All rights reserved.
#

import unittest

mldb = mldb_wrapper.wrap(mldb) # noqa

class RedditTest(MldbUnitTest):  

    def test_pipeline(self):

        mldb.put('/v1/procedures/import_reddit', { 
            "type": "import.text",  
            "params": { 
                "dataFileUrl": "http://public.mldb.ai/reddit.csv.gz",
                'delimiter':'', 
                'quoteChar':'',
                'outputDataset': 'reddit_raw',
                'limit': 2000,
                'runOnCreation': True
            } 
        })
        mldb.query("select * from reddit_raw limit 5")

        mldb.put('/v1/procedures/reddit_import', {
            "type": "transform",
            "params": {
                "inputData": """select tokenize(lineText, {offset: 1, value: 1}) as * 
                from reddit_raw""",
                "outputDataset": "reddit_dataset",
                "runOnCreation": True
            }
        })
        mldb.query("select * from reddit_dataset limit 5")


        mldb.put('/v1/procedures/reddit_svd', {
            "type" : "svd.train",
            "params" : {
                "trainingData" : """
                    SELECT 
                        COLUMN EXPR (AS columnName() ORDER BY rowCount() DESC, 
                            columnName() LIMIT 4000) 
                    FROM reddit_dataset
                """,
                "columnOutputDataset" : {
                    "id": "reddit_svd_embedding", 
                    "type":"embedding",
                    "params": {"metric":"euclidean"}
                },
                "rowOutputDataset" : {
                    "id": "reddit_svd_embedding_rows", 
                    "type":"embedding",
                    "params": {"metric":"cosine"} #test that cosine doesn't crash
                },
                "modelFileUrl": "file://tmp/MLDB-466.svd",
                "functionName": "embedder",
                "runOnCreation": True
            }
        })
        mldb.query("select * from reddit_svd_embedding limit 5")

        mldb.query("""
            select rowName(), rowPath(), rowPathElement(0) 
            from reddit_svd_embedding 
            limit 5
            """)
        
        mldb.query("""
            select embedder({ row: {HongKong: 1} })
            """)

        mldb.put("/v1/functions/nearest_subreddit", {
            "type": "embedding.neighbors",
            "params": {
                "dataset": "reddit_svd_embedding",
                "defaultNumNeighbors": 1,
                "columnName": "embedding"
            }
        })

        mldb.log("querying for nearest subreddit")
        
        result = mldb.get("/v1/query", format="aos", q="""
            select 
                embedder({ row: {soccer: 1} })[embedding] as soccer,
                neighbour.* as *, 
                x.distance
            from transpose(( 
                select nearest_subreddit({ coords: 
                    embedder({ row: {soccer: 1} })[embedding]
                })[distances] as *
                named 'distance'
            )) as x
            join reddit_svd_embedding as neighbour
            on (x.rowName() = neighbour.rowName())
        """).json()[0]

        #result now contains a vector and its nearest neighbour
        #plus distance from the nearest-neighbour function
        #let's independently reconfirm that the distance is correct

        dist_accum = 0
        for i in range(100):
            diff = result["embedding.%d" % i]-result["soccer.%d" % i]
            dist_accum += diff*diff

        import math
        # MLDB-1677
        self.assertLess(abs(math.sqrt(dist_accum) - result["x.distance"]), 0.00001)

        mldb.put('/v1/procedures/reddit_kmeans', {
            "type" : "kmeans.train",
            "params" : {
                "trainingData" : "select * from reddit_svd_embedding",
                "outputDataset" : "reddit_kmeans_clusters",
                "numClusters" : 20,
                "runOnCreation": True
            }
        })
        mldb.query("select * from reddit_kmeans_clusters limit 5")


        mldb.put('/v1/procedures/reddit_tsne', {
            "type" : "tsne.train",
            "params" : {
                "trainingData" : "select * from reddit_svd_embedding",
                "rowOutputDataset" : "reddit_tsne_embedding",
                "runOnCreation": True
            }
        })
        mldb.query("select * from reddit_tsne_embedding limit 5")


        mldb.put('/v1/procedures/reddit_count_users', {
            "type": "transform",
            "params": {
                "inputData": """select columnCount() as numUsers 
                    from transpose(reddit_dataset)""",
                "outputDataset": "reddit_user_counts",
                "runOnCreation": True
            }
        })
        mldb.query("select * from reddit_user_counts limit 5")

        mldb.query("""
            select *, quantize(x, 7) as grid_x, quantize(y, 7) as grid_y 
            named regex_replace(rowName(), '\|1', '') 
            from merge(reddit_tsne_embedding, reddit_kmeans_clusters) as r
                join reddit_user_counts as c on c.rowName() = r.rowPathElement(0)
            where cluster is not null 
            order by numUsers desc
        """)

mldb.run_tests()
